File: solutions/c/bob/1/bob.c
#include "bob.h"
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>

char *hey_bob(char *greeting){
 
    char* ans = (char*)malloc(300*sizeof(char));
    char * pg = (char*)malloc((strlen(greeting)+1)*sizeof(char));
    strcpy(pg,greeting);
    size_t gl=strlen(pg);

    if (all_blanks(pg)) {
        ans="Fine. Be that way!";
        return ans;
    }
    for (size_t gi=gl-1; pg[gi]==' '; gi--) {
        pg[gi]='\0';
    }
    gl=strlen(pg);

    if (!strcmp(uc_str(pg),pg)&&!no_letters(pg)) {
        if (greeting[gl-1] == '?') {
            ans="Calm down, I know what I'm doing!";
            return ans;
        } 
        ans="Whoa, chill out!";
        return ans;
    }
    if (pg[gl-1] == '?') {
        ans="Sure.";
        return ans;
    }
    ans="Whatever.";
    return ans;
}
    
char *uc_str(char* str) {
    char* uc = (char*)malloc(strlen(str));
    for (size_t i=0;i<strlen(str);i++) uc[i]=toupper(str[i]);
    return uc;
}

bool no_letters(char* str) {
    for (size_t i=0;i<strlen(str);i++) {
        if (isalpha(str[i])) return false;
    }    
    return true;
}

bool all_blanks(char* str) {
    for (size_t i=0;i<strlen(str);i++) {
        if (str[i]!=' '&&str[i]!='\t'&&str[i]!='\r'&&str[i]!='\n') return false;
    }    
    return true;
}

File: solutions/fortran/bob/1/bob.f90
module bob
  implicit none
contains

  function hey(statement)
    character(100) :: hey
    character(len=*), intent(in) :: statement
    if (index(statement,"?")>0 .and. &
        statement == toUpper(statement) .and. &
        hasLetter(statement)) then
      hey = "Calm down, I know what I'm doing!" 
    else if (index(trim(statement),"?")== &
               len_trim(statement) .and. & 
               len_trim(statement)>0) then
      hey="Sure."
    else if (statement == toUpper(statement) .and. &
             hasLetter(statement)) then
      hey="Whoa, chill out!" 
    else if (len_trim(statement)==0) then
      hey = "Fine. Be that way!"
    else 
      hey = "Whatever."
    end if 
    return
  end function hey

  function toUpper(lc)
    character(len=*), intent(in) :: lc
    character(100) :: toUpper
    integer :: i 
    toUpper=""
    do i=1,len(lc)
      if (lc(i:i)>='a' .and. lc(i:i)<='z') then
        toUpper(i:i) = achar(iachar(lc(i:i)) - 32)
      else
        toUpper(i:i) = lc(i:i)
      end if
    end do
    return
  end function toUpper
  
  logical function hasLetter(s)
    character(len=*), intent(in) :: s
    integer :: i 
    do i=1,len(s)
      if (s(i:i)>='a' .and. s(i:i)<='z' .or. &
          s(i:i)>='A' .and. s(i:i)<='Z' ) then
        hasLetter=.true.
        return
      end if
    end do
    hasLetter=.false.
    return
  end function hasLetter
end module bob
